test/jdk/java/text/Format/MessageFormat/MessageFormatToPatternTest.java

```diff
@@ -0,0 +1,323 @@
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @summary Verify MessageFormat.toPattern() is equivalent to original pattern
+ * @bug 8323699
+ * @run junit MessageFormatToPatternTest
+ */
+
+import java.text.ChoiceFormat;
+import java.text.DateFormat;
+import java.text.DecimalFormat;
+import java.text.Format;
+import java.text.MessageFormat;
+import java.text.NumberFormat;
+import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.Locale;
+import java.util.Random;
+import java.util.stream.Stream;
+
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class MessageFormatToPatternTest {
+
+    private static final int NUM_RANDOM_TEST_CASES = 1000;
+    private static final int MAX_FORMAT_NESTING = 3;
+
+    private static Locale savedLocale;
+    private static long randomSeed;             // set this to a non-zero value for reproducibility
+    private static Random random;
+    private static boolean spitSeed;
+    private static int textCount;
+
+// Setup & Teardown
+
+    @BeforeAll
+    public static void setup() {
+        savedLocale = Locale.getDefault();
+        Locale.setDefault(Locale.US);
+        if (randomSeed == 0)
+            randomSeed = new Random().nextLong();
+        random = new Random(randomSeed);
+    }
+
+    @AfterAll
+    public static void teardown() {
+        Locale.setDefault(savedLocale);
+    }
+
+// Tests
+
+    @Test
+    public void testJDK_8323699() {
+
+        // This is the test case from JDK-8323699
+        testRoundTrip(new MessageFormat("{2,choice,0.0#option A: {0}|1.0#option B: {0}'}'}"));
+
+        // A few more test cases from the PR#17416 discussion
+        testRoundTrip(new MessageFormat("Test: {0,number,foo'{'#.00}"));
+        testRoundTrip(new MessageFormat("Test: {0,number,foo'}'#.00}"));
```

Comment:
Can we add some more concrete test cases, for example, escaped quotes within a MessageFormat pattern (represented by doubled single quotes). 

Another funny case is something like a MessageFormat created by `"{0,number,' abc }'' ' 0.00}"` which becomes `"{0,number, abc '}'''  #0.00}"` after created from the value returned by toPattern() from the first MessageFormat. The Strings appear really different but format equivalently, it would be nice to have suspicious cases like these explicitly defined.

Reply:
Sure thing - added in 58e8cc68f45.